cli/azd/cmd/deploy.go

```diff
@@ -178,7 +178,7 @@ func (d *deployAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 
 		stepMessage := fmt.Sprintf("Deploying service %s", svc.Config.Name)
 		d.console.ShowSpinner(ctx, stepMessage, input.Step)
-		result, progress := svc.Deploy(ctx, d.azdCtx)
+		result, progress := svc.Deploy(ctx, d.azdCtx, env)
```

Comment:
Do we really need the entire `env` here?

The `env` is used as input arg to create the Project:
`proj, err := projConfig.GetProject(ctx, env, d.console, d.azCli, d.commandRunner)`

The Project contains the services. So, can we use the `GetProject()` function to create the service with the overridden endpoint?  Then we don't need to pass the `env` again later for each service. 

When creating the `type ServiceTarget interface` , we should use some argument to set the overridden endpoint.

Reply:
You have a good observation here regarding good object design. I've addressed the redundant passing concern, but didn't address the object construction concern. 

Fundamentally, you're asking should `Service` have full access to full environment? In the principle of least exposure approach, I agree with you. We should limit access, and also possibly moving hydration of environment -> state logic into a `NewService` constructor. In the construction, global environment state is read and local state that is relevant to the Service is set. This introduces maximal reusability of `Service`. This is the most flexible design, and we should explore this as the codebase evolves. I'm not going to address this right now since it would be inconsistent with how `ServiceTarget` and `FrameworkService` is constructed. I don't think we currently would take advantage of this flexibility (testing, composability or otherwise), so let's take a stab at it when approaching it.